core/src/block.rs

```diff
@@ -19,894 +19,635 @@ use iroha_data_model::{block::*, events::prelude::*, transaction::prelude::*};
-        }
-
-        /// Return signatures that are verified with the `hash` of this block, removing all other signatures.
-
-        #[inline]
-        pub fn retain_verified_signatures(&mut self) -> impl Iterator<Item = &SignatureOf<Self>> {
-            self.signatures.retain_verified_by_hash(self.hash())
-        }
-
-        /// Revalidate transaction hashes, verify signatures and produce [`CommittedBlock`]
-        ///
-        /// # Errors
-        /// - If transaction hashes don't match the hashes stored in the block header
-        /// - If signatures verification fails
-        pub fn revalidate(
-            mut self,
-            topology: &Topology,
-        ) -> Result<CommittedBlock, (Self, eyre::Report)> {
-            let verified_signatures = self.retain_verified_signatures();
-
-            if topology
-                .filter_signatures_by_roles(
-                    &[
-                        Role::ValidatingPeer,
-                        Role::Leader,
-                        Role::ProxyTail,
-                        Role::ObservingPeer,
-                    ],
-                    verified_signatures,
-                )
-                .len()
-                .lt(&topology.min_votes_for_commit())
-            {
-                return Err((
-                    self,
-                    eyre!("The block doesn't have enough valid signatures to be committed."),
-                ));
-            }
-
-            self.revalidate_hashes()
-        }
-
-        /// Revalidate transaction hashes and produce [`CommittedBlock`]
-        ///
-        /// When calling this function, the user is responsible for the validity of the block signatures.
-        /// Preference should be given to [`Self::revalidate`], where signature verification is built in.
-        ///
-        /// # Errors
-        /// - If transaction hashes don't match the hashes stored in the block header
-        pub fn revalidate_hashes(self) -> Result<CommittedBlock, (Self, eyre::Report)> {
-            if self
+}
+/// Revalidate transaction hashes on block.
+/// # Errors
+/// * The hashes were not valid.
+pub fn revalidate_hashes(
+    block: VersionedCommittedBlock,
+) -> Result<VersionedCommittedBlock, (VersionedCommittedBlock, eyre::Report)> {
+    match block {
+        VersionedCommittedBlock::V1(block) => {
+            if block
                 .transactions
                 .iter()
                 .map(VersionedValidTransaction::hash)
                 .collect::<MerkleTree<_>>()
                 .hash()
-                .ne(&self.header.transactions_hash)
+                .ne(&block.header.transactions_hash)
             {
-                return Err((self, eyre!("The transaction hash stored in the block header does not match the actual transaction hash.")));
+                return Err((block.into(), eyre!("The transaction hash stored in the block header does not match the actual transaction hash.")));
```

Comment:
Do you think having a more structured error (e.g. its own error type) in `block` would be beneficial? Getting rid of `eyre` in this module would also allow to get rid of `Sized` bound on `Revalidate`, though I agree it's minor.

Reply:
`eyre` should be removed from `iroha_core` and errors should be made strongly typed. `eyre` is not an appropriate error handling solution for libraries. We're slowly removing it so this might be a good time to remove it from block validation api

Reply:
I would like to keep this issue focused on the collapsing of the block types. Therefore I have created this issue concerning eyre, https://github.com/hyperledger/iroha/issues/3353